dist_example.py

```diff
@@ -2,135 +2,75 @@
-        # number)
-        n_workers = 16
-
-        # Give all the data to each worker -- let's focus on
-        # expensive computation with small data for now (it can easily be
-        # generalized).
-        batch_size = get_batch_size(model._updates, base=64, increase=0.05)
-        idx = rng.choice(len(train_set), size=batch_size)
-        worker_idxs = np.array_split(idx, n_workers)
-
-        # Distribute the computation of the gradient. In practice this will
-        # mean (say) 4 GPUs to accelerate the gradient computation. Right now
-        # for ease it's a small network that doesn't need much acceleration.
-        grads = [
-            gradient(
-                inputs,
-                targets,
-                model=deepcopy(model),
-                loss=F.nll_loss,
-                device=device,
-                idx=worker_idx,
-            )
-            for worker_idx in worker_idxs
-        ]
-
-        # The gradients have been calculated. Now, let's make sure
-        # ``optimizer`` can see the gradients.
-        optimizer.zero_grad()
-        num_data = sum(info["_num_data"] for info in grads)
-        assert num_data == batch_size
-        for name, param in model.named_parameters():
-            grad = sum(grad[name] for grad in grads)
-            param.grad = grad / num_data
-
-        # Take an optimization step
-        optimizer.step()
-
-        # From here to the end of the loop is metadata tracking: how many
-        # updates/epochs (or passes through the data), what's the loss, etc
-        model._updates += 1
-        model._num_eg_processed += num_data
-
-        epochs = model._num_eg_processed / len(train_set)
-        epochs_since_print = epochs - (_num_eg_print / len(targets))
-        loss_avg = sum(grad["_loss"] for grad in grads) / num_data
-        if _num_eg_print == -1 or epochs_since_print > verbose:
-            _num_eg_print = model._num_eg_processed
-            print(
-                "epoch={:0.3f}, updates={}, loss={:0.3f}".format(
-                    epochs, model._updates, loss_avg
-                )
-            )
-        if model._num_eg_processed >= len(targets) + _start_eg:
-            break
-    return {"num_data": model._num_eg_processed, "batch_loss": loss_avg}
-
+def _get_gradients(client, model_future, n, idx, batch_size, n_workers):
+    """
+    Calculates the gradients at a given state
+    """
+    # pass a seed here
+    rng = np.random.RandomState(seed=idx) # Deterministic way to generate random numbers
+    idx = rng.choice(n, size=batch_size) # size random indexes
+    worker_idxs = np.array_split(idx, n_workers)
+
+    ## This is a dirty hack. See https://github.com/dask/distributed/issues/3807
+    ## for more detail
+    #  train_set, test_set = _get_fashionmnist()
+    #  train_set = client.scatter(train_set, broadcast=True)
+    train_set = None
+
+    # compute gradients
+    grads = [
+        client.submit(
```

Comment:
Side note: I'm not a huge fan of `submit`. It's too declarative; `map` is more functional and cleaner:

``` python
fn = lambda worker_idx: gradient(train_set, model=model_future, loss=F.nnl_loss, idx=worker_idx)
grads = client.map(fn, worker_idxs)
```

The `lambda` keyword creates an anonymous function. It's as if we had defined `def fn(worker_idx): ...`. Here's a good overview on `lambda`: https://realpython.com/python-lambda/